Deleting all IngressLink CR resources resetting default RD of CIS-managed Partition to RD 0

#### Setup Details
 CIS Version : 2.16.1      
 Build: f5networks/k8s-bigip-ctlr:latest       
 BIGIP Version: Big IP 17.1.1.3 
 AS3 Version: 3.46.2  
 Agent Mode: AS3       
 Orchestration: OCP      
 Orchestration Version: 4.14.16
 Pool Mode: Cluster      
 Additional Setup details: Static Routes, CRD, IngressLink

#### Description  
After deleting all IngressLink CRs from the cluster, CIS would push bare AS3 declaration with no `defaultRouteDomain` parameter, resetting the Partition default RD to 0 and breaking subsequent Static Route push.

#### Steps To Reproduce
1) Create BIG-IP Partition with non-0 RD (e.g. RD1) to be managed by CIS
2) Deploy CIS in ClusterIP + Static Route + CRD + IngressLink + RD1
3) Create IngressLink CR(s)
4) Delete all IngressLink CRs
5) Observe CIS-managed Partition in BIG-IP (`tmsh list auth partition`) --> Partition will have RD0 as default RD

#### Expected Result 
Default RD for CIS-managed Partition is consistent with both CIS deployment args (i.e. `--default-route-domain`) and original Partition RD in BIG-IP

#### Actual Result  
CIS-Managed Partition has RD reset to 0 with no IngressLink CR present, due to CIS pushed bare AS3 declaration without `<tenant>/defaultRouteDomain` parameter. This makes every Static Route push fails until another CR is created.  
In the worst timing scenario, 30 seconds (or more, depends on the `verify-interval` parameter) would elapse after next CR is created with no connectivity, until the next Static Route push is accepted by BIG-IP.

#### Diagnostic Information

CIS Pod creation log

```CIS-pod-log
2024/05/05 12:25:46 [INFO] [INIT] Starting: Container Ingress Services - Version: 2.16.1, BuildInfo: azure-5932-c0934efcc07227fdc64bd9c8e17e8cc21a4bc3b2
..
2024/05/05 12:25:47 [DEBUG] [AS3] BIGIP is serving with AS3 version: 3.46.2
2024/05/05 12:25:47 [DEBUG] Controller Created
2024/05/05 12:25:47 [DEBUG] Client Created
..
2024/05/05 12:25:47 [DEBUG] Starting Custom Resource Worker
2024/05/05 12:25:47 [DEBUG] Processing Node Updates for static routes
2024/05/05 12:25:47 [DEBUG] [CCCL] ConfigWriter (0xc000194ed0) writing section name static-routes
2024/05/05 12:25:47 [DEBUG] [CCCL] ConfigWriter (0xc000194ed0) successfully wrote section (static-routes)
2024/05/05 12:25:47 [DEBUG] Wrote static route config section: {[]}
```

AS3 declaration after deleting CR
```AS3-declaration
{
  "CIS-Inglink": {
    "Shared": {
      "class": "Application",
      "template": "shared"
    },
    "class": "Tenant",
    "label": "CIS-Inglink"
  },
  "class": "ADC",
  "controls": {
    "class": "Controls",
    "userAgent": "CIS/v2.16.1 OCP/v4.14.16",
    "archiveTimestamp": "2024-05-05T12:32:13.239Z"
  },
  "id": "urn:uuid:85626792-9ee7-46bb-8fc8-4ba708cfdc1d",
  "label": "CIS Declaration",
  "remark": "Auto-generated by CIS",
  "schemaVersion": "3.46.0",
  "updateMode": "selective"
}
```  

#### Observations (if any)
Similar issue might also happen in other CR

Can now confirm that the same symptom also happens in VirtualServer and TransportServer CRs

```
2024/05/06 12:58:08 [DEBUG] [AS3] Response from BIG-IP: code: 200 --- tenant:CIS-CRD --- message: success
2024/05/06 12:58:08 [DEBUG] [AS3] Raw response from Big-IP: map[declaration:map[CIS-CRD:map[Shared:map[class:Application template:shared] class:Tenant label:CIS-CRD] class:ADC controls:map[archiveTimestamp:2024-05-06T12:57:59.069Z class:Controls userAgent:CIS/v2.16.1 OCP/v4.14.16] id:urn:uuid:85626792-9ee7-46bb-8fc8-4ba708cfdc1d label:CIS Declaration remark:Auto-generated by CIS schemaVersion:3.46.0 updateMode:selective] results:[map[code:200 host:localhost lineCount:26 message:success runTime:2552 tenant:CIS-CRD]]]
...
2024/05/06 12:58:11 [ERROR] [2024-05-06 12:58:11,032 f5_cccl.service.manager ERROR] F5CcclResourceRequestError - 400 Unexpected Error: Bad Request for uri: https://10.0.0.9:443/mgmt/tm/net/route/ Text: '{"code":400,"message":"01070330:3: Static route gateway 10.0.0.133 is not directly connected via an interface.","errorStack":[],"apiError":3}'
```

--- a/docs/RELEASE-NOTES.rst
+++ b/docs/RELEASE-NOTES.rst
@@ -8,7 +8,7 @@ Added Functionality
 ```````````````````
 **What's new:**
     * CRD
-        * `Issue 3337 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3378>`_: Support to control ciphers groups and ssl options in TLSProfile CRD
+        * `Issue 3378 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3378>`_: Support to control ciphers groups and ssl options in TLSProfile CRD
     * Improved performance for Hub Mode using the isTenantNameServiceNamespace label in the AS3 configmap, See `Example <https://github.com/F5Networks/k8s-bigip-ctlr/blob/2.x-master/docs/config_examples/configmap/user-defined-configmap/hubmode-configmap/hubmode-cmap.yaml>`_
     * Pod Graceful Shutdown support for AS3 ConfigMap using CIS deployment parameter *pod-graceful-shutdown*
 
@@ -17,6 +17,7 @@ Bug Fixes
 * `Issue 3395 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3395>`_: BIGIP controller 2.16.0 removes F5 configuration when removing Kubernetes resources in namespace.
 * `Issue 3424 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3424>`_: Static routes not added if a label not added to a namespace when --namespace-label used.
 * `Issue 3443 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3443>`_: Fix issue with IPAM IP allocation on resource recreation
+* `Issue 3406 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3406>`_: Upon deleting all the CRD resources, the default route domain of CIS-managed Partition resets to 0
 
 Upgrade notes
 ``````````````

--- a/pkg/controller/backend.go
+++ b/pkg/controller/backend.go
@@ -644,7 +644,7 @@ func (agent *Agent) createAS3LTMConfigADC(config ResourceConfigRequest) as3ADC {
 	for tenant := range agent.cachedTenantDeclMap {
 		if _, ok := config.ltmConfig[tenant]; !ok && !agent.isGTMTenant(tenant) {
 			// Remove partition
-			adc[tenant] = getDeletedTenantDeclaration(agent.Partition, tenant, cisLabel)
+			adc[tenant] = getDeletedTenantDeclaration(agent.Partition, tenant, cisLabel, config)
 		}
 	}
 	for tenantName, partitionConfig := range config.ltmConfig {
@@ -656,7 +656,7 @@ func (agent *Agent) createAS3LTMConfigADC(config ResourceConfigRequest) as3ADC {
 		partitionConfig.PriorityMutex.RUnlock()
 		if len(partitionConfig.ResourceMap) == 0 {
 			// Remove partition
-			adc[tenantName] = getDeletedTenantDeclaration(agent.Partition, tenantName, cisLabel)
+			adc[tenantName] = getDeletedTenantDeclaration(agent.Partition, tenantName, cisLabel, config)
 			continue
 		}
 		// Create Shared as3Application object
@@ -689,7 +689,7 @@ func (agent *Agent) createAS3LTMConfigADC(config ResourceConfigRequest) as3ADC {
 	return adc
 }
 
-func getDeletedTenantDeclaration(defaultPartition, tenant, cisLabel string) as3Tenant {
+func getDeletedTenantDeclaration(defaultPartition, tenant, cisLabel string, config ResourceConfigRequest) as3Tenant {
 	if defaultPartition == tenant {
 		// Flush Partition contents
 		sharedApp := as3Application{}
@@ -698,6 +698,7 @@ func getDeletedTenantDeclaration(defaultPartition, tenant, cisLabel string) as3T
 		return as3Tenant{
 			"class":              "Tenant",
 			as3SharedApplication: sharedApp,
+			"defaultRouteDomain": config.defaultRouteDomain,
 			"label":              cisLabel,
 		}
 	}

--- a/pkg/controller/backend.go
+++ b/pkg/controller/backend.go
@@ -644,7 +644,7 @@ func (agent *Agent) createAS3LTMConfigADC(config ResourceConfigRequest) as3ADC {
 	for tenant := range agent.cachedTenantDeclMap {
 		if _, ok := config.ltmConfig[tenant]; !ok && !agent.isGTMTenant(tenant) {
 			// Remove partition
-			adc[tenant] = getDeletedTenantDeclaration(agent.Partition, tenant, cisLabel, config)
+			adc[tenant] = getDeletedTenantDeclaration(agent.Partition, tenant, cisLabel, &config)
 		}
 	}
 	for tenantName, partitionConfig := range config.ltmConfig {
@@ -656,7 +656,7 @@ func (agent *Agent) createAS3LTMConfigADC(config ResourceConfigRequest) as3ADC {
 		partitionConfig.PriorityMutex.RUnlock()
 		if len(partitionConfig.ResourceMap) == 0 {
 			// Remove partition
-			adc[tenantName] = getDeletedTenantDeclaration(agent.Partition, tenantName, cisLabel, config)
+			adc[tenantName] = getDeletedTenantDeclaration(agent.Partition, tenantName, cisLabel, &config)
 			continue
 		}
 		// Create Shared as3Application object
@@ -689,7 +689,7 @@ func (agent *Agent) createAS3LTMConfigADC(config ResourceConfigRequest) as3ADC {
 	return adc
 }
 
-func getDeletedTenantDeclaration(defaultPartition, tenant, cisLabel string, config ResourceConfigRequest) as3Tenant {
+func getDeletedTenantDeclaration(defaultPartition, tenant, cisLabel string, config *ResourceConfigRequest) as3Tenant {
 	if defaultPartition == tenant {
 		// Flush Partition contents
 		sharedApp := as3Application{}

--- a/pkg/controller/backend_test.go
+++ b/pkg/controller/backend_test.go
@@ -481,9 +481,9 @@ var _ = Describe("Backend Tests", func() {
 				defaultRouteDomain: 1,
 			}
 			cisLabel := "test"
-			deletedPartition := getDeletedTenantDeclaration("test", "test", cisLabel, config)
+			deletedPartition := getDeletedTenantDeclaration("test", "test", cisLabel, &config)
 			Expect(deletedPartition[as3SharedApplication]).NotTo(BeNil())
-			deletedPartition = getDeletedTenantDeclaration("test", "default", cisLabel, config)
+			deletedPartition = getDeletedTenantDeclaration("test", "default", cisLabel, &config)
 			Expect(deletedPartition[as3SharedApplication]).To(BeNil())
 		})
 	})

--- a/pkg/controller/backend_test.go
+++ b/pkg/controller/backend_test.go
@@ -474,10 +474,16 @@ var _ = Describe("Backend Tests", func() {
 			Expect(val).NotTo(BeNil())
 		})
 		It("Test Deleted Partition", func() {
+			config := ResourceConfigRequest{
+				ltmConfig:          make(LTMConfig),
+				shareNodes:         true,
+				gtmConfig:          GTMConfig{},
+				defaultRouteDomain: 1,
+			}
 			cisLabel := "test"
-			deletedPartition := getDeletedTenantDeclaration("test", "test", cisLabel)
+			deletedPartition := getDeletedTenantDeclaration("test", "test", cisLabel, config)
 			Expect(deletedPartition[as3SharedApplication]).NotTo(BeNil())
-			deletedPartition = getDeletedTenantDeclaration("test", "default", cisLabel)
+			deletedPartition = getDeletedTenantDeclaration("test", "default", cisLabel, config)
 			Expect(deletedPartition[as3SharedApplication]).To(BeNil())
 		})
 	})

